README.md

```diff
@@ -0,0 +1,400 @@
+
+```rs
+struct TypedMetadata<T: ?Sized>(pub <T as core::ptr::Pointee>::Metadata);
+
+
+impl<T, U> CoerceUnsized<TypedMetadata<U>> for TypedMetadata<T>
+where
+    T: ?Sized + Unsize<U>,
+    U: ?Sized,
+{
+    fn coerce_unsized(self) -> TypedMetadata<U> {
+        TypedMetadata(Unsize::target_metadata(self.0))
+    }
+}
+```
+
+## Pin Unsoundness
+
+See the following issue for context: https://github.com/rust-lang/rust/issues/68015
+
+The design of the new traits here do not address the underlying issue in regards to `Pin`.
+The author of this RFC feels like addressing the `Pin` soundness in the definitions of the traits is wrong, as in almost all cases where are a user implements one of these traits `Pin` will be irrelevant to them. And at its core, unsizing is not coupled with `Pin` whatsoever.
+It would make much more sense to fix the unsound `CoerceUnsized` implementation that `Pin` provides.
+That is given the current implementation of (with the new definition of the trait):
+```rs
+
+// Copied from core library docs:
+// Note: this means that any impl of `CoerceUnsized` that allows coercing from
+// a type that impls `Deref<Target=impl !Unpin>` to a type that impls
+// `Deref<Target=Unpin>` is unsound. Any such impl would probably be unsound
+// for other reasons, though, so we just need to take care not to allow such
+// impls to land in std.
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    // U: core::ops::Deref, this bound is accidentally missing upstream, hence we can't use `Pin::new_unchecked` in the implementation
+{
+    fn coerce_unsized(self) -> Pin<U> {
+        Pin {
+            pointer: self.pointer.coerce_unsized(),
+        }
+    }
+}
+```
+
+We should rather strife to have the following 2 implementations:
+```rs
+// Permit going from Pin<impl Unpin> to Pin<impl Unpin>
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    P: Deref<Target: Unpin>,
+    U: Deref<Target: Unpin>,
+{
+    fn coerce_unsized(self) -> Pin<U> {
+        Pin::new(self.pointer.coerce_unsized())
+    }
+}
+// Permit going from Pin<impl Pin> to Pin<impl Pin>
+impl<P, U> CoerceUnsized<Pin<U>> for Pin<P>
+where
+    P: CoerceUnsized<U>,
+    P: core::ops::Deref<Target: !Unpin>,
+    U: core::ops::Deref<Target: !Unpin>,
+{
+    fn coerce_unsized(self) -> Pin<U> {
+        // SAFETY: The new unpin Pin is derived from another unpin Pin, so we the pinned contract is kept up
+        unsafe { Pin::new_unchecked(self.pointer.coerce_unsized()) }
+    }
+}
+```
+
+
+While this is a breaking change, it should be in line with being a soundness fix.
+Unfortunately, these kind of impl requires negative bounds and negative reasoning which is its own can of worms and therefor likely not to happen, see https://github.com/rust-lang/rust/issues/42721.
+Though maybe allowing them for auto traits alone could work out fine, given those types of traits are already rather special.
```

Comment:
Does this mean this RFC would be blocked on negative bounds? If yes, can the pin part be broken out of this RFC and be done at a later point? If yes, we should clarify that.

Reply:
Well it depends. Assuming the negative bounds would be "accepted" as the fix, then its not blocked on them. Because this RFC does not change the status quo in regards to the unsoundness (it doesn't make it worse). I'll add that as a note.